### DISCUSSION TITLE
Oberserved rounding effects in `cyrb53a`

### DISCUSSION THREAD

**Discussion by goto40** [bryc/code#22] [Q&A] [answered]
Thank you for sharing your great work!

I observed that we have a small rounding effect in the last multiplication of `cyrb53a`:
Replacing 
```js
  return 2097152 * (h2 >>> 0) + (h1 >>> 11);;
```
with
```js
  return BigInt(2097152) * BigInt((h2 >>> 0) + (h1 >>> 11));
```
yields slightly different values.

I apologize if I have overseen this in the documentation, but I thought the algorithm was designed in order not to have rounding issues due to the precision of the `number` type. Of course this has no effect for most use cases - but if you want to check against a reference implementation (e.g. in C) you have to deal with this...

**Comment by bryc** [comment, marked as answer]
You're doing `(2097152) * (h2 + h1)` when it should be `(2097152) * (h2) + (h1)`.

The multiply is doing a bitwise left-shift, i.e. `h2 << 21`. We can't use the `<<` operator because bitwise operations are bound to 32 bits, which prevents us from outputting a 53-bit `Number` as intended.

Keep in mind, `BigInt` has some performance overhead, which can have an accumulative effect on large-scale simulations, and honestly you might as well use all 64 bits using `4294967296n * BigInt(h2) + BigInt(h1)`.

The 53-bit transform is just an alternative way of using `h1` and `h2` (which is kind of required in JS); C users can (and should) just simply do `(h2 << 32) | h1`. The way to verify implementations would be to simply check `h1` and `h2`, not any reduction/transformation.

Also I don't recommend using `cyrb53a` in anything serious because:

a) It's not finished. When it is finished, it will produce different results, so if you expect your cyrb53a function to be correct, you have to plan to update it in the future, or you'll have to just continue using the 'beta' version.
b) `cyrb53` is not *bad*. The non-uniformity (aka quality of randomness) that was found in vanilla `cyrb53` is very minor (and requires a lot of statistical math to even observe!), and has no real impact on collision occurrences. `cyrb53a` is currently only a slight improvement in this area, and further research must be done to confirm the results, and refine it even further.
